Add support for CUDA

nweston opened the pull request:
I noticed that `bear` does not add CUDA source files to `compile_commands.json`. Clang-based tools can understand CUDA (e.g. I use rtags for completion in `.cu` files), so it's useful to have these in the database.

I was able to fix this by recognizing `nvcc` as a C++ compiler and classifying `.cu` files as C++ sources.

I'm happy to add a test for this but I don't recognize the test system that you're using so I may need some help with that.

rizsotto commented:
Thanks @nweston , this looks good.
I'm wondering, what tool will benefit from having the CUDA files in the compilation database? Or generally, does a CUDA project mix C++ and CUDA source files? (In that case will the a `clang-tidy` or `clang-format` happy to see a `.cu` file?)

nweston commented:
I personally use `clang-format` and [rtags](https://github.com/Andersbakken/rtags) regularly on `.cu` files without any problems. CUDA is basically C++ with a few minor syntax extensions, and `clang` is actually a fully-functioning CUDA compiler these days so I would expect most clang-based tools to handle it pretty well.

rizsotto commented:
Thanks for your quick reply.
It sounds easy then. I would like to ask you to add test to this PR. And I can guide you in that...
the tests are written for [lit](https://llvm.org/docs/CommandGuide/lit.html) to execute and located in `test/functional/cases/`. (I would put the new one into `test/functional/cases/intercept/cuda`.) The `fts` files are shell scripts. If you would execute them, they would create a small project (usually empty source files and a shell script as a build file). And in the comment the lines with `RUN` will build the project and verifies the output.
The harder part is to make it run for travis-ci. Need to find a package for OSX and for Ubuntu which contains the CUDA compiler.

nweston commented:
Thanks, I've added the test.

As for running the test in Travis, do you need a real `nvcc` or could you get by with a fake? It seems that bear only looks at the names of commands and source files, so a script named `nvcc` that does nothing would generate the same compilation database as the real thing.

I tried that out in my test script and it seems to work on my machine... we'll see how it goes in Travis. 

If it's a bad idea for some reason I'm not thinking of, I can back out that commit.

rizsotto commented:
Thanks @nweston to working on this.

I see benefit to run a real CUDA compiler on Travis. (Sometimes compilers are linked statically, or have 32bit support, etc.) Would be more comforting to know that it actually works with real compiler. (You are completely right about that it's just a string match inside the code. That would be a unit test. These tests more like integration tests.)

There is package for Ubuntu xenial (called `nvidia-cuda-toolkit`). And we might need to run this test conditionally, depend on the `nvcc` available or not. (That can be achieved to add `nvcc` to the `REQUIRES` line. And add the compiler availability check to `test/lit.cfg`.)

nweston commented:
I added `nvidia-cuda-toolkit` to the Travis environment and implemented the feature check. An interesting thing happens in the CI build -- the build script runs `/usr/bin/nvcc`, but that in turns appears to run `/usr/lib/nvidia-cuda-toolkit/bin/nvcc`. bear picks up both of them, resulting in two entries for each file instead of one.

This doesn't happen with a CUDA toolkit installed directly from NVIDIA, which is what I have on my machine.

Is there a way for bear to capture the `/usr/bin/nvcc` processes but ignore the `/usr/lib/nvidia-cuda-toolkit/bin/nvcc` child processes?

rizsotto commented:
I'm happy that the test caught this case. :)

To solve the duplicated entries issue, it needs a few other changes... Bear does not read the execution files in the same order as they were created. But it runs a duplication check, (which includes the compiler name as well and not only the flags).

I might make those changes independently from this PR. (Might merge that change first to see it cause no harm and can put this change on top of it.) The only problem is that I'm travelling for work from tomorrow and won't have time.

nweston commented:
That sounds fine. Just let me know if you need any further changes from me.

rizsotto commented:
Ok, will ping you how it goes. And thanks again to work on this issue.

rizsotto commented:
An update: I made these changes (order the execution traces in time and ignore the compiler when compare), but still fails the tests... Now it contains only one entry for a CUDA file compilation (`nvcc` or `/usr/lib/nvidia-cuda-toolkit/bin/nvcc`). The problem lays in the ordering. The precision of the file creation time is depend on the file system implementation, but it usually has seconds granularity. (Putting the PID into the filename will not help, because the wrappers usually do not `fork()`, but `exec*()`. And the execution of the wrapper is really fast.)

Was thinking to make the output diff tool to ignore the compiler for this test.

rizsotto commented:
I've merged to `devel` branch, will be release as `2.4.3`.

Thanks again to do this work!
